// @slack/web-api/dist/response/SearchAllResponse.js
"use strict";
/* eslint-disable */
/////////////////////////////////////////////////////////////////////////////////////////
//                                                                                     //
// !!! DO NOT EDIT THIS FILE !!!                                                       //
//                                                                                     //
// This file is auto-generated by scripts/generate-web-api-types.sh in the repository. //
// Please refer to the script code to learn how to update the source data.             //
//                                                                                     //
/////////////////////////////////////////////////////////////////////////////////////////
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ts = exports.UserTeam = exports.LastEditor = void 0;
var LastEditor;
(function (LastEditor) {
    LastEditor["Empty"] = "";
    LastEditor["U00000000"] = "U00000000";
})(LastEditor = exports.LastEditor || (exports.LastEditor = {}));
var UserTeam;
(function (UserTeam) {
    UserTeam["Empty"] = "";
    UserTeam["T00000000"] = "T00000000";
})(UserTeam = exports.UserTeam || (exports.UserTeam = {}));
var Ts;
(function (Ts) {
    Ts["Empty"] = "";
    Ts["The0000000000000000"] = "0000000000.000000";
})(Ts = exports.Ts || (exports.Ts = {}));
